File: 347_Top_K_Frequent_Elements/solution.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        /*
            Inputs: vector of integer nums, integer k
            Output: vector of integers, the k most frequent element in any order

            We are guaranteed that the answer is unique

            sample test case
            1.) nums = [1,1,1,2,2,3], k = 2, output = [1,2]
                1 appears 3 times, 2 appears twice, 3 appears once, 2 most frequent elements are 1 and 2
            
            Strategy
            1.) Sort the array, get linear count and push it into max heap with frequency as the key
                - pop k times and push it into result vector.
                - TC: O(nlogn) from sorting O(n) for heap
            
            2.) Hashmap + maxHeap
                - obtain counts in linear time with a hashmap key will be the element and value is the count
                - push entries into maxHeap with count as key
                TC: O(n) time, O(n) space
                - worst case O(nlogn) time, O(n) space

                std::vector<tuple<int,int>>max_heap;
                std::unordered_map<int,int>count_map;
                std::vector<int>res;
                int count;
                int val;

                for (int elem:nums) {
                    count_map[elem]++;
                }

                for (auto &it:count_map) {
                    max_heap.push_back(std::make_tuple(it.second,it.first));
                }
                std::make_heap(max_heap.begin(),max_heap.end());

                for (int i=0; i<k; i++) {
                    std::pop_heap(max_heap.begin(),max_heap.end());
                    std::tie(count,val) = max_heap.back();
                    res.push_back(val);
                    max_heap.pop_back();
                }
                return res;
            
            3.) Hoare's selection algorithm O(N) avg, worst case O(n2)

            4.) Bucket sort trick
                - We know given nums, there are at most nums.size() different elements and at the opposite end, 
                    if all elements are the same, then that give at most a frequency of nums.size() for 
                    a particular element
                - We can initialize a bucket vector of vector of size nums.size()+1 where indices represent the
                frequency the elements appear. By doing so, we can iterate from the back of the bucket for the k most
                frequent element

        */
        std::unordered_map<int,int>count_map;
        std::vector<std::vector<int>> bucket(nums.size()+1);
        std::unordered_set<int>res_set;
        std::vector<int>res;
        for (int num:nums) {
            count_map[num]++;
        }
        for (auto &it:count_map) {
            bucket[it.second].push_back(it.first);
            
        }
        int i = nums.size();
        while (bucket[i].empty()) {
            --i;
        }
        
        while(k>0){
            for (int j = 0; j < bucket[i].size(); j++){
                
                if (res_set.find(bucket[i][j]) == res_set.end()) {
                    res_set.insert(bucket[i][j]);
                    res.push_back(bucket[i][j]);
                    --k;   
                }
            }
            --i;
        }
        return res;
    }
};

File: 347_Top_K_Frequent_Elements/solution.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        /*
            Input: integer array nums, integer k
            Output: intger array the k most frequent elements

            Sample test case
            1.) nums = [1,1,1,2,2,3], k = 2
                - output = [1,2], 1 and 2 are the 2 most frequent elements in nums

            2.) nums = [1], k = 1
                - output = [1]
            
            We are guaranteed the answer is unique, min length nums is 1

            Strategy
            1.) heap/priority queue ordering in terms of counts TC: O(nlogn), SC: O(n)
                - pop the k most frequent elements, push into array
                - return
            
            Follow up: TC better than O(nlogn). TC: O(n), SC: O(n)
            - What options are available if we don't want to sort conventionally?
            - bucket sort
                - hashmap to count elements
                - let the count be the bucket index, we populate the elements with the count in the buckets
                - let the other buckets be empty. Initial bucket of size number of elements in array (worst case)
                - iterate from the back of the bucket and find valid k numbers
        */
        List<Integer>[] bucket = new List[nums.length +1];
        HashMap<Integer,Integer>count_map = new HashMap<Integer,Integer>();
        List<Integer> res = new ArrayList<>();
        int idx=0;
        for (int num: nums){
            count_map.put(num,count_map.getOrDefault(num,0)+1);
        }
        for(int key: count_map.keySet()){
            
            if (bucket[count_map.get(key)] == null){
                bucket[count_map.get(key)] = new ArrayList<>();    
            }
            bucket[count_map.get(key)].add(key);
        }

        for (int i=bucket.length-1; i>=0 && res.size()<k; i--){
            if (bucket[i] != null) {
                res.addAll(bucket[i]);
            }
        }
        return res.stream().mapToInt(i->i).toArray();
    }
}
